### ISSUE TITLE
Do common_iterator's copy/move ctors/operators need to be specified?

### ISSUE THREAD

**Issue by ericniebler** [ericniebler/stl2#250]
The members of `common_iterator` are worded to permit a union-like implementation. I think we need to say what copying/assigning these things do. Also: what do we say about `constexpr`/`noexcept`?

Proposed Resolution
-------------------
(Wording relative to N4671. This wording also resolves #436 "common_iterator's destructor should not be specified in [common.iter.op=]".)

Strike the destructor declaration from the synposis of class `common_iterator` in [common.iterator] as follows:
```diff
 [...]
 common_iterator& operator=(const common_iterator<ConvertibleTo<I>, ConvertibleTo<S>>& u);
-~common_iterator();
 see below operator*();
 [...]
```

Strike [common.iterator]/1 that begins "Note: It is unspecified whether common_iterator’s members iter and sentinel ..."

Change [common.iter.op.const] as follows:

```diff
   common_iterator();

 1 Effects: Constructs a common_iterator, value-initializing is_sentinel
+  , sentinel,
   and iter. Iterator operations applied to the resulting iterator have defined behavior if and
   only if the corresponding operations are defined on a value-initialized iterator of type I.
-2 Remarks: It is unspecified whether any initialization is performed for sentinel.

   common_iterator(I i);

 3 Effects: Constructs a common_iterator, initializing is_sentinel with false
-  and
+  ,
   iter with i
+  , and value-initializing sentinel
   .

-4 Remarks: It is unspecified whether any initialization is performed for sentinel.

   common_iterator(S s);

 5 Effects: Constructs a common_iterator, initializing is_sentinel with true and sentinel with s
+  , and value-initializing iter
   .

-6 Remarks: It is unspecified whether any initialization is performed for iter.

   common_iterator(const common_iterator<ConvertibleTo<I>, ConvertibleTo<S>>& u);

 7 Effects: Constructs a common_iterator, initializing is_sentinel with u.is_sentinel
+  , iter with u.iter, and sentinel with u.sentinel
   .

-(7.1) — If u.is_sentinel is true, sentinel is initialized with u.sentinel.
-(7.2) — If u.is_sentinel is false, iter is initialized with u.iter.
-
-8 Remarks:
-(8.1) — If u.is_sentinel is true, it is unspecified whether any initialization is
-        performed for iter.
-(8.2) — If u.is_sentinel is false, it is unspecified whether any initialization is
-        performed for sentinel.
```

Change [common.iter.op=] as follows:
```diff
 common_iterator& operator=(const common_iterator<ConvertibleTo<I>, ConvertibleTo<S>>& u);

 1 Effects: Assigns u.is_sentinel to is_sentinel
+  , u.iter to iter, and u.sentinel to sentinel
   .

-(1.1) — If u.is_sentinel is true, assigns u.sentinel to sentinel.
-
-(1.2) — If u.is_sentinel is false, assigns u.iter to iter.
-
-Remarks:
-
-(1.3) — If u.is_sentinel is true, it is unspecified whether any operation is performed on iter.
-
-(1.4) — If u.is_sentinel is false, it is unspecified whether any operation is performed on
-        sentinel.

 2 Returns: *this

-  ~common_iterator();
-
-3 Effects: Destroys all members that are currently initialized.
```

**Comment by ericniebler** [comment]
Also, `common_iterator::operator=` is assigning to (possibly) uninitialized field.

**Comment by ericniebler** [comment]
@CaseyCarter for your review.

**Comment by CaseyCarter** [comment]
For the IS, I would be concerned that "unspecified whether any initialization/operation is performed on `XXX`" is underspecified - "any" operation could cover a lot of ground. I do think this is good enough to allow implementer latitude for now, and we'll almost certainly reword this in terms of `variant` for the IS anyway, so I'm happy.

**Comment by CaseyCarter** [comment]
WIP resolution:

(Wording relative to N4671. This wording also resolves #436 "common_iterator's destructor should not be specified in [common.iter.op=]".)

Strike [common.iterator]/1 that begins "Note: It is unspecified whether common_iterator’s members iter and sentinel ..."

Change [common.iter.op.const] as follows:

```diff
   common_iterator();

 1 Effects: Constructs a common_iterator, value-initializing is_sentinel
+ , sentinel,
   and iter. Iterator operations applied to the resulting iterator have defined behavior if and
   only if the corresponding operations are defined on a value-initialized iterator of type I.
-2 Remarks: It is unspecified whether any initialization is performed for sentinel.

   common_iterator(I i);

 3 Effects: Constructs a common_iterator, initializing is_sentinel with false
-  and
+  ,
   iter with i
+  , and value-initializing sentinel
   .

-4 Remarks: It is unspecified whether any initialization is performed for sentinel.

   common_iterator(S s);

 5 Effects: Constructs a common_iterator, initializing is_sentinel with true and sentinel with s
+  , and value-initializing iter
   .

-6 Remarks: It is unspecified whether any initialization is performed for iter.

   common_iterator(const common_iterator<ConvertibleTo<I>, ConvertibleTo<S>>& u);

 7 Effects: Constructs a common_iterator, initializing is_sentinel with u.is_sentinel
+  , iter with u.iter, and sentinel with u.sentinel
   .

-(7.1) — If u.is_sentinel is true, sentinel is initialized with u.sentinel.
-(7.2) — If u.is_sentinel is false, iter is initialized with u.iter.
-
-8 Remarks:
-(8.1) — If u.is_sentinel is true, it is unspecified whether any initialization is
-        performed for iter.
-(8.2) — If u.is_sentinel is false, it is unspecified whether any initialization is
-        performed for sentinel.
```

Change [common.iter.op=] as follows:
```diff
 common_iterator& operator=(const common_iterator<ConvertibleTo<I>, ConvertibleTo<S>>& u);

 1 Effects: Assigns u.is_sentinel to is_sentinel
+  , u.iter to iter, and u.sentinel to sentinel
   .

-(1.1) — If u.is_sentinel is true, assigns u.sentinel to sentinel.
-
-(1.2) — If u.is_sentinel is false, assigns u.iter to iter.
-
-Remarks:
-
-(1.3) — If u.is_sentinel is true, it is unspecified whether any operation is performed on iter.
-
-(1.4) — If u.is_sentinel is false, it is unspecified whether any operation is performed on
-        sentinel.

 2 Returns: *this

-  ~common_iterator();
-
-3 Effects: Destroys all members that are currently initialized.
```
